README.md
# Summary
This cookbook should be used as an example of how to automate the process of
running multiple instances of Microgateway on a single VM.  This doc
outlines how to accomplish this with Chef 12.0, Node.js and Nginx.

# edgemicro Cookbook

This cookbook starts Apigee Edgemicro 2.1.1 via npm and it will install the following components:
- Nginx
- Node LTS
- Edgemicro

If you enable Nginx from the attributes/default.rb file then it will
- install Nginx which will listen on port 80
- install and enable firewalld and only open ports 22, 80


## Requirements

This cookbook was tested on CentOS 7.

You must have an Apigee Edge public account.

You must create an edgemicro aware proxy.  Please see the following link which describes how to create the proxy in Edge.
- http://docs.apigee.com/microgateway/latest/edge-microgateway-tutorial

You should also have a target service to which the Apigee edgemicro proxy will send requests.

### Platforms

- CentOS 7

### Chef

- Chef 12.0 or later

### Microgateway
Microgateway is installed via the Node.js package manager, NPM, which is the preferred
way to install the Microgateway.

i.e.
```
npm install -g edgemicro@latest
```
or
```
npm install -g edgemicro@2.1.1
```

### Cookbooks
Depends on the following cookbooks when you enable nginx:
- `firewalld` to enable specific ports
- `nginx` to proxy requests to Edgemicro

Depends on the following cookbooks whether you enable or disable nginx:
- `yum` - to update all the yum repositories
- `tar` - to untar installation files
- `zipfile` - to unzip the microgateway zip file

## Attributes
- node['edgemicro']['user'] = 'vagrant'
- node['edgemicro']['initscript'] = 'init.sh'
- node['edgemicro']['initscript_mode'] = '0777'
- node['edgemicro']['version'] = '2.1.1'
- node['edgemicro']['folder_prefix'] = 'apigee-edge-micro-'
- node['edgemicro']['port'] = 8000
- node['edgemicro']['nginx_upstream_port'] = 8000
- node['edgemicro']['cluster_enabled'] = false
- node['edgemicro']['cluster_processes'] = 2
- node['edgemicro']['processes'] = 2
- node['edgemicro']['node_version'] = 'v4.4.7'
- node['edgemicro']['node_os_version'] = 'linux-x64'
- node['edgemicro']['node_tarball_mode'] = '0777'
- node['edgemicro']['nginx_enabled'] = true
- node['edgemicro']['nginx_conf'] = 'nginx.conf'
- node['edgemicro']['nginx_conf_mode'] = '0777'

Please update the following attributes
- node['edge']['org'] = 'apigee_edge_orgname'
- node['edge']['env'] = 'apigee_edge_environment'
- node['edge']['org_admin'] = 'apigee_edge_org_admin'
- node['edge']['org_admin_password'] = 'apigee_edge_org_admin_password'


<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['edgemicro']['user']</tt></td>
    <td>String</td>
    <td>edgemicro user</td>
    <td><tt>edgemicro</tt></td>
  </tr>
   <tr>
    <td><tt>['edgemicro']['init']</tt></td>
    <td>String</td>
    <td>inital run script; currently this file is blank.</td>
    <td><tt>init.sh</tt></td>
  </tr>
   <tr>
    <td><tt>['edgemicro']['initscript_mode']</tt></td>
    <td>String</td>
    <td>This is sets the mode for the init script when it is copied to the VM.</td>
    <td><tt>0777</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['version']</tt></td>
    <td>String</td>
    <td>This should be the latest version.</td>
    <td><tt>2.1.1</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['folder_prefix']</tt></td>
    <td>String</td>
    <td>The version above is appended to the folder prefix and this is the folder name that is extracted from the zip file downloaded from Apigee.</td>
    <td><tt>apigee-edge-micro-</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['port']</tt></td>
    <td>Number</td>
    <td>Default starting port for Edgemicro. Do not change this port value.</td>
    <td><tt>8000</tt></td>
  </tr>

  <tr>
    <td><tt>['edgemicro']['nginx_upstream_port']</tt></td>
    <td>Number</td>
    <td>This is a bit of a hack.  For some reason the port attribute above was not being set correctly when the Chef script executed for nginx.  So I had to define another starting port for Nginx. This should always be the same as t he port attribute above.</td>
    <td><tt>8000</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['cluster_enabled']</tt></td>
    <td>Boolean</td>
    <td>Microgateway has the ability to run in cluster mode with multiple child processes.  This enables cluster mode.</td>
    <td><tt>false</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['cluster_processes']</tt></td>
    <td>Number</td>
    <td>If you enable cluster mode, then this controls how many child processes will be started.</td>
    <td><tt>2</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['processes']</tt></td>
    <td>Number</td>
    <td>You have the ability to run multiple Edgemicro parent processes on the same VM. This attribute controls how many processes will be started.  In order to run multiple parent processes on the same VM, then each process must listen on a different port.  The starting port is 8000 and it increments that number by 1 for each additional parent process.</td>
    <td><tt>2</tt></td>
  </tr>

  <tr>
    <td><tt>['edgemicro']['node_version']</tt></td>
    <td>String</td>
    <td>This is the version of Node that Chef will install. The recommended version for Edgemicro is Node LTS.</td>
    <td><tt>v4.4.7</tt></td>
  </tr>
   <tr>
    <td><tt>['edgemicro']['node_os_version']</tt></td>
    <td>String</td>
    <td>OS version on which Node will be installed. </td>
    <td><tt>linux-x64</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['node_tarball_mode']</tt></td>
    <td>String</td>
    <td>This is the mode of the node tar.gz that is downloaded from nodejs.org.</td>
    <td><tt>0777</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['nginx_enabled']</tt></td>
    <td>Boolean</td>
    <td>Enables nginx and firewalld. Only port 80 and 22 will be opened on firewall port.  Nginx will listen on port 80 and forward requests to localhost:edgemicro_port.</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['nginx_conf']</tt></td>
    <td>String</td>
    <td>This override the default nginx.conf with our settings.</td>
    <td><tt>nginx.conf</tt></td>
  </tr>
<tr>
    <td><tt>['edgemicro']['nginx_conf_mode']</tt></td>
    <td>String</td>
    <td>This sets the mode of nginx.conf file when it is copied to the VM.</td>
    <td><tt>0777</tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['org']</tt></td>
    <td>String</td>
    <td>Apigee Edge organization name.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['env']</tt></td>
    <td>String</td>
    <td>Apigee Edge environment name.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['org_admin']</tt></td>
    <td>String</td>
    <td>Apigee Edge org admin email address.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['edgemicro']['org_admin_password']</tt></td>
    <td>String</td>
    <td>Apigee Edge org admin password.</td>
    <td><tt></tt></td>
  </tr>
</table>


## Usage

Make sure to update the following attributes in the attributes/default.rb file.
- node['edge']['org'] = 'apigee_edge_orgname'
- node['edge']['env'] = 'apigee_edge_environment'
- node['edge']['org_admin'] = 'apigee_edge_org_admin'
- node['edge']['org_admin_password'] = 'apigee_edge_org_admin_password'

### Enable Nginx and Firewalld
This will enable Nginx and enable Firewalld on the CentOS machine.
node['edgemicro']['nginx_enabled'] = true

### Change how many parent processes to start
The following attribute will determine how many parent Edgemicro processes to start.

e.g. The processes attribute is set to 2. Therefore, Chef will start 2 Edgemicro processes the first one listening on port 8000 and the second one on listening on port 8001.
- node['edgemicro']['port'] = 8000
- node['edgemicro']['processes'] = 2

If you change the node['edgemicro']['processes'] to 4, then it will start 4 parent processes starting at port 8000 and ending on port 8003.  These ports must be available on the VM.

### Enable Edgemicro cluster mode
By default Edgemicro cluster is disabled.
- node['edgemicro']['cluster_enabled'] = false

To enable Microgateway cluster mode then change the values as shown below.
- node['edgemicro']['cluster_enabled'] = true
- node['edgemicro']['cluster_processes'] = 2

With the setting above, cluster mode is enabled.  The cluster_proceses attribute determines how many child processes each Edgemicro gateway parent process will create.

e.g. If cluster mode is enabled and the cluster_processes attribute is set to 2, then one Edgemicro process will start 2 child processes.  Therefore, there will be 3 processes all together (1 parent and 2 children).


### Kitchen
cd into the edgemicro directory:

- `kitchen list` to list all the kitchen recipes.
- `kitchen converge default-centos-7` to start edgemicro running in an virtual box machine.
- `kitchen destroy` to destroy the edgemicro virtual box VM.
- `kitchen login` to login to the VM.

#### Make sure you change the IP address
The Kitchen configuration uses Vagrant as the driver and Vagrant only allows
bridged networking.  Therefore, you must change the IP address to a unique IP
in the same network range as the hosts interface IP.

```
suites:
  - name: default
    driver:
      vm_hostname: edgemicro.com
      network:
      - ["public_network", {ip: "192.168.103.110", bridge: "en0: Wi-Fi (AirPort)"}]
    run_list:
      - recipe[edgemicro]
    attributes:
```

For example, if you run `ifconfig` in your terminal on a Mac or `ipconfig` on
Windows, then you should see your IP address.

```
en0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
  ...
  inet 192.168.111.101 netmask ... broadcast 192.168.111.255
	...
	status: active
```

Copy your IP address and change the last octet to a different value - `192.168.111.102`.
Paste it into the `edgemicro/.kitchen.yaml` file.

execute
`kitchen destroy`

then
`kitchen converge`

#### Error received during start up
If you receive an error similar to the one below. Then it means that Vagrant could not
determine which network interface to use to assign your machine a public IP.

```
==> default: Available bridged network interfaces:
1) en0: Wi-Fi (AirPort)
2) en1: Thunderbolt 1
3) en2: Thunderbolt 2
4) p2p0
5) awdl0
6) bridge0
```

In this case you need to copy the first item below and paste it in the
edgemicro/.kitchen.yaml file. Replace the value next to bridge with the
first item from the list displayed on your machine.  

```
suites:
  - name: default
    driver:
      vm_hostname: edgemicro.com
      network:
      - ["public_network", {ip: "192.168.103.110", bridge: "en0: Wi-Fi (AirPort)"}]
    run_list:
      - recipe[edgemicro]
    attributes:
```


#### After kitchen up finishes execution
Once Kitchen finished executing you should see a message similar to the one below.

```
-----> Kitchen is finished. (3m5.22s)
```

Now you can run the following command to login to the VM and see nginx and edgemicro running.
```
kitchen login
```

Grep for edgemicro.  You can see the edgemicro instances running on their respective ports.
```
# ps -ef | grep edgemicro

root     12303     1  1 19:09 ?        00:00:00 node /usr/local/bin/edgemicro start -o org -e env -k key -s secret --port 8000
root     12318     1  1 19:09 ?        00:00:00 node /usr/local/bin/edgemicro start -o org -e env -k key -s secret --port 8001
root     12380 12366  0 19:10 pts/0    00:00:00 grep --color=auto edgemicro
```

Grep for nginx. You can see Nginx is running.
```
#ps -ef | grep nginx

root     12232     1  0 19:08 ?        00:00:00 nginx: master process nginx -c /home/vagrant/nginx/nginx.conf
nginx    12233 12232  0 19:08 ?        00:00:00 nginx: worker process
root     12386 12366  0 19:11 pts/0    00:00:00 grep --color=auto nginx
```
#### Send request to nginx
Now you can send requests to the nginx server, which will route requests to the edgemicro.
We start Nginx listening on port 80 by default.

Get an access token.
```
curl -i -X POST --user client_id:secret "http://<org>-<env>.apigee.net/edgemicro-auth/token" -d '{"grant_type": "client_credentials"}' -H "Content-Type: application/json"
```

```
curl -i -H "Authorization: Bearer bearer_token" http://{ip_of_vm}:80/base_path/pathsuffix
```

### Berksfile
cd into the edgemicro directory.
- `berks install` to install the dependent cookbooks to your local machine (~/.berkshelf).


### edgemicro::default

Just include `edgemicro` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[edgemicro]"
  ]
}
```


## Contributing
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

## License and Authors

Authors: Sean Williams

Apache 2.0


CHANGELOG.md
# edgemicro CHANGELOG

This file is used to list changes made in each version of the edgemicro cookbook.

## 0.2.2
- Sean Williams - Initial release of edgemicro

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.

## 0.2.3
- Updated project so that it pulls edgemicro@2.1.1 via npm.


recipes/default.rb
#
# Cookbook Name:: edgemicro
# Recipe:: default
#
# Copyright 2016, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#

include_recipe 'yum'
include_recipe 'tar'
include_recipe 'firewalld'
#include_recipe 'yum-epel'
#include_recipe 'net-tools'

user_home = '/home/' + node['edgemicro']['user']

user 'Creating edgemicro user' do
  username node['edgemicro']['user']
  comment 'edgemicro user'
  home "#{user_home}"
  shell '/bin/bash'
end


cookbook_file 'Create init script in users home directory' do
  path "#{user_home}/" + node['edgemicro']['initscript']
  user node['edgemicro']['user']
  source node['edgemicro']['initscript']
  mode node['edgemicro']['initscript_mode']
end

execute node['edgemicro']['initscript'] do
  command './' + node['edgemicro']['initscript']
  #cwd '/root'
  #user 'root'
  cwd "#{user_home}"
  user node['edgemicro']['user']
end

if (node['edgemicro']['nginx_enabled'] == true) then
  include_recipe 'nginx'

  ##open port 80
  firewalld_service 'http' do
    action :add
    zone   'public'
  end

  directory "#{user_home}/nginx" do
  #directory '/home/vagrant/nginx' do
    #path "#{user_home}/nginx"
    #path "/home/vagrant/nginx"
    owner node['edgemicro']['user']
    group node['edgemicro']['user']
    mode '0777'
    action :create
  end

  #Update the ngnix conf file
  cookbook_file "#{user_home}/nginx/" + node['edgemicro']['nginx_conf'] do
  #cookbook_file "/home/vagrant/nginx/" + node['nginx']['conf'] do
    owner node['edgemicro']['user']
    path "#{user_home}/nginx/" + node['edgemicro']['nginx_conf']
    #path "/home/vagrant/nginx/" + node['nginx']['conf']
    source node['edgemicro']['nginx_conf']
    mode node['edgemicro']['nginx_conf_mode']
  end


  replacement_string = ''
  port = node['edgemicro']['nginx_upstream_port']

  ruby_block 'update the replacement_string' do
    block do
      #Chef::Config.from_file("/etc/chef/client.rb")
      for i in 1..node['edgemicro']['processes'] do
        if (i > 1)
          port += 1
          #previous_port = port_current - 1
        end
        replacement_string += 'server localhost:' + port.to_s + ";\r\n"
      end

    end
    action :run
  end


  bash 'replacement string' do
    code <<-EOF
      echo "#{replacement_string}"
    EOF
  end

  ruby_block 'update the nginx.conf file' do
    block do

      #Find and replace the contents of nginx.conf
      file_names = ["#{user_home}/nginx/" + node['edgemicro']['nginx_conf']]
      #file_names = ["/vagrant/home/nginx/" + node['nginx']['conf']]

      file_names.each do |file_name|
        text = File.read(file_name)
        new_contents = text.gsub(/REPLACEME/, "#{replacement_string}")

        # To merely print the contents of the file, use:
        puts new_contents

        # To write changes to the file, use:
        File.open(file_name, "w") {|file| file.puts new_contents }
      end
    end
    action :run
  end

  nginx_conf = node['edgemicro']['nginx_conf']


  #restart nginx gracefully
  bash 'Restart nginx' do
    code <<-EOF
      nginx -s stop
      nginx -c #{user_home}/nginx/#{nginx_conf}
    EOF
  end

end #if (node['edgemicro']['nginx_enabled'] == true)

node_tar_ball = 'node-' + node['edgemicro']['node_version'] + '-' + node['edgemicro']['node_os_version'] + '.tar.gz'
node_uri = 'https://nodejs.org/dist/' + node['edgemicro']['node_version'] + "/#{node_tar_ball}"

=begin
remote_file "Download Node LTS" do
  path "#{user_home}/#{node_tar_ball}"
  source "#{node_uri}"
  owner node['edgemicro']['user']
  group node['edgemicro']['user']
  mode node['node']['tarball_mode']
  action :create
  #notifies :run, 'execute[untar]', :immediate
end
=end

tar_extract "#{node_uri}" do
  target_dir '/usr/local/'
  creates '/usr/local/bin/node'
  tar_flags [ '-P','--strip-components 1' ]
  #user node['edgemicro']['user']
  #user 'root'
end

#install edgemicro via npm
bash 'Install edgemicro via npm' do
  code <<-EOF
    npm install -g edgemicro@#{node['edgemicro']['version']}
  EOF
end


results = "/tmp/configure_output.txt"
file results do
  action :delete
end

#ENV['MICROGATEWAY_INSTALL'] = "/usr/local/share/" + node['edgemicro']['folder_prefix'] + node['edgemicro']['version']
ENV['EDGEMICRO_KEY'] = ''
ENV['EDGEMICRO_SECRET'] = ''

node_results = "/tmp/node_version.txt"
file node_results do
  action :delete
end

bash 'configure microgateway' do
  #user node['edgemicro']['user']
  code <<-EOF
    node -v > /tmp/node_version.txt
    edgemicro init &> init.txt
    edgemicro configure -o #{node['edge']['org']} -e #{node['edge']['env']} -u #{node['edge']['org_admin']} -p #{node['edge']['org_admin_password']} &> #{results}
  EOF
end

#echo $MICROGATEWAY_INSTALL
#$MICROGATEWAY_INSTALL/cli/edgemicro configure -o org -e env -u user@apigee.com -p password &> #{results}

=begin
execute "configure microgateway" do
  cwd "/apigee-edge-micro-2.0.4/cli"
  command "./edgemicro configure -o org -e env -u user@apigee.com -p password &> #{results}"
end
=end

key = "/tmp/key.txt"
file key do
  action :delete
end

secret = "/tmp/secret.txt"
file secret do
  action :delete
end

bash 'extract key and secret' do
  code <<-EOF
  grep key: /tmp/configure_output.txt | sed 's/key:\s//' | sed -e 's/^[ \t]*//' > /tmp/key.txt
  grep secret: /tmp/configure_output.txt | sed 's/secret:\s//' | sed -e 's/^[ \t]*//' > /tmp/secret.txt
EOF
#environment ({ 'EDGEMICRO_KEY' => "/tmp/key.txt" })
#environment ({ 'EDGEMICRO_SECRET' => "/tmp/secret.txt" })
end

bash 'extract key and secret' do
  code <<-EOF
EOF
environment ({ 'EDGEMICRO_KEY' => "cat /tmp/key.txt" })
environment ({ 'EDGEMICRO_SECRET' => "cat /tmp/secret.txt" })
end

port_current = node['edgemicro']['port']
#directory = "#{user_home}"
directory = "root"


for i in 1..node['edgemicro']['processes'] do
  if (i > 1)
    port_current += 1
    previous_port = port_current - 1

=begin
    #this code is not needed anymore since we have the --port option
    bash 'change port' do
        #user node['edgemicro']['user']
        #action :nothing
        #notifies :run, 'bash[reset port back to 8000]', :immediate
        code <<-EOF
          sudo sed -i -- 's/#{previous_port}/#{port_current}/g' #{directory}/.edgemicro/#{node['edge']['org']}-#{node['edge']['env']}-config.yaml
          sudo cat #{directory}/.edgemicro/#{node['edge']['org']}-#{node['edge']['env']}-config.yaml
        EOF
    end
=end
  end

  if (node['edgemicro']['nginx_enabled'] == false) then
    firewalld_port port_current.to_s + '/tcp' do
      action :add
      zone   'public'
    end
  end

  start_results = "/tmp/start_output_port_" + port_current.to_s + ".txt"
  file start_results do
    action :delete
  end

  if node['edgemicro']['cluster_enabled'] == true
    bash 'start microgateway as background process in cluster mode ' + port_current.to_s do
      #user node['edgemicro']['user']
      #notifies :run, 'bash[start 2nd microgateway as background process]', :immediate
      #user 'vagrant' #this does not run the command as the vagrant user
      code <<-EOF
        edgemicro start -o #{node['edge']['org']} -e #{node['edge']['env']} -k $(cat /tmp/key.txt) -s $(cat /tmp/secret.txt) -c -p #{node['edgemicro']['cluster_processes']} --port #{port_current} &> #{start_results} &
        sleep 7
      EOF
    end

  else

    bash 'start microgateway as background process ' + port_current.to_s do
      #user node['edgemicro']['user']
      #notifies :run, 'bash[start 2nd microgateway as background process]', :immediate
      #user 'vagrant' #this does not run the command as the vagrant user
      code <<-EOF
        edgemicro start -o #{node['edge']['org']} -e #{node['edge']['env']} -k $(cat /tmp/key.txt) -s $(cat /tmp/secret.txt) --port #{port_current} &> #{start_results} &
        sleep 7
      EOF
    end
  end
end
